### DISCUSSION TITLE
Nx 16.7.4: how to configure dependency check to find root level package.json

### DISCUSSION THREAD

**Discussion by hrajauria** [nrwl/nx#19403] [General]
**Current Behavior**

I upgraded my nx version to 16.7.4 and created a few libraries which throw an error for missing dependencies due to the @nx-dependency-check.

**Description**

I have a monorepo setup with a single package.json on the workspace level. All older libraries/apps do not have a package.json and refer to the common one in the root of the package.

I created new libraries using @nx/js:library plugin and selected tsc as the bundler. As a new feature, it defaults to creating a package.json per library if it is buildable.  [Link to documentation on this.](https://nx.dev/nx-api/eslint-plugin/documents/dependency-checks#usage)

The lint target to these libraries fail since it checks the library's package.json for dependencies.

**How can I tell nx that the package.json is maintained on the workspace level while creating the libs or skip on generating the package.json at all in the new nx version?**

**Comment by loxy** [comment]
Hi!

This is also something I don't understand.

In [docs for dependency-management](https://nx.dev/concepts/more-concepts/dependency-management) it is stated that there are two ways of maintaining deps:

- Independently Maintained Dependencies
- Single Version Policy

The recommendation is to use Single Version Policy, which is also again stated in multiple issues: (https://github.com/nrwl/nx/issues/3023 or https://github.com/nrwl/nx/issues/1777)

But when I create a buildable lib, a `package.json` is created. How does that fit together? Especially in combination with ESLint rule `@nx/dependency-checks`, which is also applied always.

I also found this (which is somehow related):
- https://github.com/nrwl/nx/issues/15360
- https://github.com/nrwl/nx/issues/19989
- https://github.com/nrwl/nx/issues/9510
- https://github.com/nrwl/nx/issues/20451
- https://github.com/nrwl/nx/issues/20070
- https://github.com/nrwl/nx/issues/20029

My main questions are:

- Do I always have to specify dependencies in my buildable libs even if I use Single Version Policy?
  - If yes: Should I specify all of my deps as peer deps and using `*` as version?
-  Can/should I remove that ESLint rule when I use Single Version Policy?

The thread is now open for nearly two month. Can someone from the core team please enlighten us? Thank you!

**Comment by tbarlow12** [comment]
Having the same issue here - any updates?

**Reply by luisalfonsosandoval** [reply]
I'm also having the same issue

**Comment by fgirardey** [comment]
I have the same issue as well, I don't want to put every packages I use in libs into their respectives `package.json` files because it would break the same version policy. Is there an opened issue for that?

**Comment by binaryartifex** [comment]
if this could be looked at that'd be awesome. creating workspace plugins? they come prebaked with their own package.json. @nx/remix apps? come with their own package.json. whats the point of an integrated monorepo if you cant keep it....integrated?

**Comment by patrickmichalina** [comment]
Does this help? (check the overrides property)

```js
{
    files: ['package.json'], // Only for these files
    rules: {
      '@nx/dependency-checks': 'off', // Disable it for root-level JSON files
    },
  },
```

**Comment by moatorres** [comment]
AFAIK, `@nx/dependency-checks` will warn you about version mismatches and rectify them, therefore upholding the single version policy. An even better solution is to use [Syncpack](https://jamiemason.github.io/syncpack/) as it can also sort properties to keep `package.json` files consistent. [I'm due to write an article about this setup](https://moatorres-blog.vercel.app/blog/technology/breaking-down-libraries-in-nx-monorepos-with-typescript-project-references#whats-next-part-3). Here's a non-exhaustive configuration (`.syncpackrc`) :
```json
{
  "dependencyTypes": [
    "prod",
    "peer",
    "dev"
  ],
  "sortFirst": [
    "name",
    "description",
    "version",
    "scripts",
    "engines",
    "license",
    "type",
    "main",
    "types",
    "module",
    "exports"
  ],
  "semverGroups": [
    {
      "label": "Allow minor or patch level version updates for dependencies",
      "dependencyTypes": [
        "prod",
        "dev"
      ],
      "range": "^"
    }
  ],
  "versionGroups": [
    {
      "label": "Use workspace protocol when developing local packages",
      "dependencies": [
        "$LOCAL"
      ],
      "dependencyTypes": [
        "dev"
      ],
      "pinVersion": "workspace:*"
    },
    {
      "label": "@types packages should only be under devDependencies",
      "dependencies": [
        "@types/**"
      ],
      "dependencyTypes": [
        "!dev",
        "!peer"
      ],
      "isBanned": true
    }
  ]
}
```

**Reply by moatorres** [reply]
For the record, there's further discussion on the subject in [this thread](https://discord.com/channels/1143497901675401286/1365596517351559228). For the sake of TL;DR, I quote myself:
>Keeping all dependencies in the root `package.json` is not a good practice. How are you supposed to know which packages use X and Y dependencies further down the road when you have hundreds or thousands? Everyone will be afraid to remove a dependency and the workspace starts getting bloated. Not to mention potential branch conflicts over lock files and invalidating the whole workspace to install one dependency. IMHO, you should aim at having self-contained libraries as much as possible.
